src/Symfony/Bundle/FrameworkBundle/Tests/Functional/SecurityTest.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Tests\Functional;
+
+use Symfony\Component\Security\Core\User\User;
+
+class SecurityTest extends AbstractWebTestCase
+{
+    /**
+     * @dataProvider getUsers
+     */
+    public function testLoginUser(string $username, array $roles, ?string $firewallContext)
+    {
+        $user = new User($username, 'the-password', $roles);
+        $client = $this->createClient(['test_case' => 'Security', 'root_config' => 'config.yml']);
+
+        if (null === $firewallContext) {
+            $client->loginUser($user);
+        } else {
+            $client->loginUser($user, $firewallContext);
+        }
+
+        $client->request('GET', '/'.($firewallContext ?? 'main').'/user_profile');
+        $this->assertEquals('Welcome '.$username.'!', $client->getResponse()->getContent());
+    }
+
+    public function getUsers()
+    {
+        yield ['the-username', ['ROLE_FOO'], null];
+        yield ['the-username', ['ROLE_FOO'], 'main'];
+        yield ['other-username', ['ROLE_FOO'], 'custom'];
+
+        yield ['the-username', ['ROLE_FOO'], null];
+        yield ['no-role-username', [], null];
+    }
+
+    public function testLoginUserMultipleRequests()
+    {
+        $user = new User('the-username', 'the-password', ['ROLE_FOO']);
+        $client = $this->createClient(['test_case' => 'Security', 'root_config' => 'config.yml']);
+        $client->loginUser($user);
+
+        $client->request('GET', '/main/user_profile');
+        $this->assertEquals('Welcome the-username!', $client->getResponse()->getContent());
+
+        $client->request('GET', '/main/user_profile');
+        $this->assertEquals('Welcome the-username!', $client->getResponse()->getContent());
+    }
+
+    public function testLoginInBetweenRequests()
+    {
+        $user = new User('the-username', 'the-password', ['ROLE_FOO']);
+        $client = $this->createClient(['test_case' => 'Security', 'root_config' => 'config.yml']);
+
+        $client->request('GET', '/main/user_profile');
+        $this->assertTrue($client->getResponse()->isRedirect('http://localhost/login'));
+
+        $client->loginUser($user);
+
+        $client->request('GET', '/main/user_profile');
+        $this->assertEquals('Welcome the-username!', $client->getResponse()->getContent());
+    }
```

Comment:
What about adding some tests to check for users that are **not** in the provider?

Reply:
It has to be in the user provider (as Symfony reloads the user upon every requests and if the user changed (i.e. is no longer available), the user is logged out). Do you mean adding a test that the request is unsuccesfull?